File: 314-binary-tree-vertical-order-traversal/binary-tree-vertical-order-traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var verticalOrder = function(root) {
     if (!root) return [];

  const map = new Map();

  function dfs(node, column, row) {
    if (!node) return;

    if (!map.has(column)) map.set(column, []);
    // Store both row and value to sort later
    map.get(column).push({ row, val: node.val });

    // Traverse left and right subtrees
    dfs(node.left, column - 1, row + 1);
    dfs(node.right, column + 1, row + 1);
  }

  dfs(root, 0, 0);

  const sortedColumns = Array.from(map.keys()).sort((a, b) => a - b)
  console.log(sortedColumns)
  const levels = [];

  for (const col of sortedColumns) {
    // Sort nodes in each column by their row index
    const sortedNodes = map.get(col).sort((a, b) => a.row - b.row).map(item => item.val);
    levels.push(sortedNodes);
  }

  return levels;

};

File: 314-binary-tree-vertical-order-traversal/binary-tree-vertical-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def verticalOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []

        column_map = defaultdict(list)

        queue = deque([(root, 0)])
        min_col = max_col = 0

        while queue:
            node, column = queue.popleft()

            column_map[column].append(node.val)

            if node.left:
                queue.append((node.left, column - 1))
                if column - 1 < min_col:
                    min_col = column - 1
            if node.right:
                queue.append((node.right, column + 1))
                if column + 1 > max_col:
                    max_col = column + 1

        return [column_map[i] for i in range(min_col, max_col + 1)]